src/charm.py

```diff
@@ -242,6 +232,81 @@ def restart_mongos_service(self) -> None:
         self.stop_mongos_service()
         self.start_mongos_service()
 
+    def share_uri(self) -> None:
+        """Future PR - generate URI and give it to related app"""
+        # TODO future PR - generate different URI for data-integrator as that charm will not
+        # communicate to mongos via the Unix Domain Socket.
+
+    def set_user_role(self, roles: List[str]):
```

Comment:
this function will be called by `mongos` charm when it gets notified by the hosting `application` charm about which database / user roles the `application` charm needs

Reply:
> Typically App -><- MongoS relation gets a relation-changed event, where it turns out that the App came up with a new set of roles.
Then MongoS receives the event, gets the data (i.e. new set of roles), and calls this function to "register" the new set of roles on the OTHER relation, which is MongoS -><- Mongo?

This exactly is the proposed workflow. 


> The reason why I'm asking, is because I'd like to know that noone can directly call this function (i.e. implicitly modify the MongoS -><- Mongo relation) but MongoS itself -- from an event handler.

Yes, this is true. the only "thing" that would call this function is `mongos` from the event `mongos` recieves from `application` about the creation of roles/users (we might support changing of roles/users, but that is a topic for a separate conversation)